src/Pact/Eval.hs

```diff
@@ -1226,15 +1228,85 @@ reduceDirect r _ ai = evalError ai $ "Unexpected non-native direct ref: " <> pre
 
 initPact :: Info -> PactContinuation -> Term Ref -> Eval e (Term Name)
 initPact i app bod = view eePactStep >>= \es -> case es of
-  Just v -> evalError i $ "initPact: internal error: step already in environment: " <> pretty v
-  Nothing -> view eeHash >>= \hsh ->
-    applyPact i app bod $ PactStep 0 False (toPactId hsh) Nothing
+  Just v@(PactStep step b (PactId parent) _) -> do
+    whenExecutionFlagSet FlagDisablePact43 $
+      evalError i $ "initPact: internal error: step already in environment: " <> pretty v
+    let Hash name' = pactHash $ T.encodeUtf8 $ renderCompactText (_pcDef app)
+        newPactId = toPactId (pactHash (T.encodeUtf8 parent <> ":" <> name'))
+    applyNestedPact i app bod $ PactStep step b newPactId Nothing
```

Comment:
How is a subsequently-executed pact prevented from being treated as a nested? It looks like I can issue at top-level `(pact1 foo bar) (pact2 baz quux)` and `pact2` will be created as a nested child of `pact1`.

Reply:
```
 Nothing -> view eeHash >>= \hsh -> do
    let pStep = PactStep 0 False (toPactId hsh) Nothing
    local (set eePactStep (Just pStep)) $ applyPact i app bod pStep mempty
```
`eePactStep` is set as `local`. Two top level pacts won't be nested since the call to  `pact1` returns a result and the pactstep is no longer set.